### ISSUE TITLE
OCaml 5 performance regression for unmarshal-heavy workloads

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#13300]
Today @OlivierNicole, @ejgallego and myself worked on understanding the performance of Coq on OCaml 5 -- some benchmarks suggested that it is 10-20% slower than on OCaml 4 on average, see https://github.com/coq/coq/pull/19225# .

TL;DR: the OCaml 5.2 implementation of `intern.c` is less efficient than in OCaml 4 for good/subtle reasons, but there also seems to be a GC-pacing change on top of this which hurts the Coq Require/Import performance noticeably.

This is an issue, not a PR, because we are not sure what is a good fix for this problem.

## On the Coq side

We observed a clear regression in the performance of Coq Require/Import statements, which seem to mostly exercise the unmarshalling logic of the OCaml runtime. For example on a file from the coq CoLoR library, the important statements are 33% slower with OCaml 5. (Imports are typically a small constant cost at the beginning of each Coq .v file, a few seconds at most; but many projects have a lot of small .v files, so for some projects import costs take a substantial fraction of the total time.)

The input file that we use for Coq performance testing is

```coq
From Coq Require Import FSets Structures.OrderedType.
From CoLoR Require Import LogicUtil BoolUtil VecUtil FSetUtil NatUtil RelUtil
     SN SetUtil.
From CoLoR Require Union.
```

(This comes from the `coq-color` project, it is the very beginning of the file LTerm.v -- we looked at this file because the Coq benchmark infrastructure reported a ~15% slowdown for the whole file. To test this on your machine it suffices to install the `coq-color` package in a switch, so the dependencies will be installed and that code above can be built with `coqc`.)

Compiling this code with `coqc` seems to open 303 .vo files, which weight 30Mio in total.

```
$ strace coqc -w -all Lterm_imports.v 2>&1 | grep openat | grep -o "/home/gasche.*.vo"  | wc -l
302
```

On OCaml 4, the top of the `perf` profile looks like this:

```
  17.67%  coqc     coqc                    [.] intern_rec
   6.41%  coqc     coqc                    [.] caml_oldify_mopup
   5.40%  coqc     coqc                    [.] caml_oldify_one
   5.01%  coqc     coqc                    [.] Stdlib.Map.update_572
```

On OCaml 5, it looks like this:

```
  28.91%  coqc     coqc                    [.] do_some_marking
  11.55%  coqc     coqc                    [.] caml_shared_try_alloc
   8.97%  coqc     coqc                    [.] intern_rec
   4.12%  coqc     coqc                    [.] Stdlib.Map.update_580
   3.52%  coqc     coqc                    [.] oldify_one
   3.30%  coqc     coqc                    [.] oldify_mopup
```

We believe that `do_some_marking` and `caml_shared_try_alloc` represent extra costs that were not present in OCaml 4, causing the observed slowndown.

### OCaml 4

On OCaml 4, the marshaller allocates a single block of memory, writes all the unmarshalled blocks into it sequentially, and adds it to the memory owned by the major heap.

https://github.com/ocaml/ocaml/blob/8eb41f72ded84df884c3671734c947f612091f84/runtime/intern.c#L803-L805

https://github.com/ocaml/ocaml/blob/8eb41f72ded84df884c3671734c947f612091f84/runtime/intern.c#L375-L376

https://github.com/ocaml/ocaml/blob/8eb41f72ded84df884c3671734c947f612091f84/runtime/intern.c#L694-L698

(Later some of these blocks will become free, and they will be freed by the major GC machinery.)

### OCaml 5.2: lots of `caml_shared_try_alloc`

On OCaml 5.2, this strategy cannot be used, because the major heap is stored in size-segregated pools (a pool for all blocks of size 2, another for blocks of size 3, etc.), and there is no support for storing objects of different sizes contiguously on the major heap.

Instead we have to call `caml_shared_try_alloc` on each of the marshalled objects -- except in the corner case where the marshalled data fits in the minor heap, where the OCaml 4 approach is used. See https://github.com/ocaml-multicore/ocaml-multicore/pull/768 for more details.

https://github.com/ocaml/ocaml/blob/74a76cc2f2588eb18a443a60918f040838292ac7/runtime/intern.c#L499

https://github.com/ocaml/ocaml/blob/74a76cc2f2588eb18a443a60918f040838292ac7/runtime/intern.c#L431-L433

Note: intern.c thus allocates objects directly in the major heap, instead of trying to put small objects in the minor heap. This is a sensible heuristic because unmarshalled objects typically live long, longer than the next minor GC.

(Note to self: we could change the *marshaller* to group objects per size in the marshalled format, so that we could directly unmarshal objects into the right pool size.)

 This explains why `caml_shared_try_alloc` shows up so high in the profile above -- but not the `do_some_marking` cost, which is even larger, as `caml_try_shared_alloc` (or the rest of the unmarshaller) does not cause any major GC slice / any marking work.

I *think* (not sure yet) that the extra `caml_shared_try_alloc` calls are not causing the performance regression, they are mostly shifting costs from `intern_rec` to this function. My hypothesis is that the slowdown comes from extra marking work that is performed by the major GC.
 
### GC pacing

Whenever `intern.c` allocates an object (directly in the major heap), it increases some local counter `d->allocated_words`, which tracks the number of words allocated into the major heap.

This counter is also increased when promoting objects from the minor heap, or when we allocate an object larger than `Max_young_wosize`, directly in the major heap. In those other cases, there is a check and, if `allocated_words` becomes higher than some limit, a major GC slice is requested:

https://github.com/ocaml/ocaml/blob/74a76cc2f2588eb18a443a60918f040838292ac7/runtime/memory.c#L434-L438

This was already the case with OCaml 4, but for OCaml 4 the limit to request a GC slice was `minor_heap_wsz` and not `minor_heap_wsz / 5`:

https://github.com/ocaml/ocaml/blob/8eb41f72ded84df884c3671734c947f612091f84/runtime/memory.c#L494-L498

The change from `heap_wsz` to `heap_wsz / 5` was made in OCaml 5, in one of the various shady GC-pacing changes by @damiendoligez: #11903 . We believe that this causes many additional major GC slices requested by `intern`-heavy workflows, and cause the abnormal `do_some_marking` time observed in our profile.

### An experiment / quick hack

We don't know of a clean fix for this issue in 5.2, but we know of a non-clean fix: if we *remove* the change to `allocated_words` in `intern.c` (we stop counting unmarshalled blocks to decide to request new major slices), then the Import/Require machinery of Coq suddenly becomes about 40% faster than OCaml 5.2, and ends up slightly faster than OCaml 4.

```
Summary
  ".../modified-5.2/coqc -w -all Lterm_imports.v" ran
    1.06 ± 0.03 times faster than .../vanilla-4.14/coqc -w -all Lterm_imports.v
    1.40 ± 0.04 times faster than .../vanilla-5.2/coqc -w -all Lterm_imports.v
```

The diff:
```diff
diff --git a/runtime/intern.c b/runtime/intern.c
index 2be12e3479e..a767fd50adf 100644
--- a/runtime/intern.c
+++ b/runtime/intern.c
@@ -430,7 +430,7 @@ static value intern_alloc_obj(struct caml_intern_state* s, caml_domain_state* d,
   } else {
     p = caml_shared_try_alloc(d->shared_heap, wosize, tag,
                               0 /* no reserved bits */);
-    d->allocated_words += Whsize_wosize(wosize);
+    //d->allocated_words += Whsize_wosize(wosize);
     if (p == NULL) {
       intern_cleanup (s);
       caml_raise_out_of_memory();
```

### What would be a proper fix?

A few months ago @damiendoligez and @stedolan noticed that the use of `heap_wsz / 5` causes various performance regressions for *promotion*-heavy workflows, and introduced in trunk (#13086) a distinction between `allocated_words` (all words that end up in the major heap) and `allocated_words_direct` (directly in the major heap). Currently (in trunk) the `intern.c` code increments both `allocated_words` and `allocated_words_direct`, so we believe that trunk still suffers from the issue we described.

We think that the following approaches could work:
- not increment `allocated_words_direct` from the marshaller anymore (this mirrors our own hack)
- or, increment `allocated_words_direct`, but only for unmarshalled blocks that are larger than `Max_young_wosize`; in essence, the suggestion is that maybe `allocated_words_direct` should be thought of as counting the number of words above Max_young_wosize, and not the number of non-promoted words.

(We haven't tested these ideas as Coq does not currently build on trunk.)

**Comment by gasche** [comment]
Note: the input file that we use for Coq performance testing is

```coq
From Coq Require Import FSets Structures.OrderedType.
From CoLoR Require Import LogicUtil BoolUtil VecUtil FSetUtil NatUtil RelUtil
     SN SetUtil.
From CoLoR Require Union.
```

Compiling this code with `coqc` seems to open 303 .vo files, which weight 30Mio in total.

```
$ strace coqc -w -all Lterm_imports.v 2>&1 | grep openat | grep -o "/home/gasche.*.vo"  | wc -l
302
```

**Comment by kayceesrk** [comment]
Thanks for the investigation and the writeup. 

> Currently (in trunk) the intern.c code increments both allocated_words and allocated_words_direct, so we believe that trunk still suffers from the issue we described.

Is the next step to confirm that it is indeed the case? It may be easiest to backport #13086 to 5.2 given that https://github.com/coq/coq/pull/19225 does not build on trunk.

**Comment by gasche** [comment]
I tried to create a synthetic benchmark that reproduces the performance difference, but this harder than I expected: many workflows of the form "do a mix of unmarshalling and unrelated work" have the same speed with 4.14 and 5.2, and some of them go faster with 5.2 -- and this is sensitive to GC settings, etc. For now I don't understand the specifics of the Require/Import workload that explain the observed slowdown.

**Comment by xavierleroy** [comment]
Thanks for the detailed analysis.  Allocation of multiple blocks (as opposed to OCaml 4's "allocate one big block and carve sub-blocks inside" strategy) is probably unavoidable in OCaml 5, but your figures suggest that it is not that costly.  Besides, the OCaml 4 strategy was suspected of causing fragmentation in the major heap.  GC pacing is an issue, though.

**Comment by kayceesrk** [comment]
Do you have reproduction instructions for the benchmark? I'm unable to install `coq-color` with coq from https://github.com/coq/coq/pull/19225.

**Comment by gasche** [comment]
I have the following in my notes:

```
opam switch create coq-ocaml5-debug 5.2.0 --repos=default,coq-released=https://coq.inria.fr/opam/released
eval $(opam env --switch=coq-ocaml5-debug --set-switch)
opam install coq-color
```
(Building the coq-color dependencies, including Coq in particular, probably takes around 15-20 minutes.)

This installs coq 8.19.2, which is what we used for our experiments yesterday.

**Comment by OlivierNicole** [comment]
As for me I was able to reproduce on a coq source tree with the following sequence of commands:

```
opam source --dev-repo coq
mv coq coq-ocaml5
cd coq-ocaml5
make -j8 world
make -j8 ci-color
cd ..
# Run and profiling of coqc on a test .v file
cat <<EOF > TestMarshal.v
From Coq Require Import FSets Structures.OrderedType.
From CoLoR Require Import LogicUtil BoolUtil VecUtil FSetUtil NatUtil RelUtil
     SN SetUtil.
From CoLoR Require Union.
EOF
dune exec --root coq-ocaml5 -- perf record --call-graph lbr -- coq-ocaml5/_build/default/dev/shim/coqc -w -all -R coq-ocaml5/_build_ci/color/ CoLoR TestMarshal.v
```

**Comment by kayceesrk** [comment]
I see quite a different result on my MacBook with M2 chip.

```bash
% hyperfine -r 20 --warmup 5 '~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v' '~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v' '~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v'
Benchmark 1: ~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     554.5 ms ±  38.3 ms    [User: 505.5 ms, System: 41.2 ms]
  Range (min … max):   469.8 ms … 623.8 ms    20 runs
 
Benchmark 2: ~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     387.0 ms ±  10.1 ms    [User: 342.3 ms, System: 38.8 ms]
  Range (min … max):   379.6 ms … 425.6 ms    20 runs
 
Benchmark 3: ~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     402.5 ms ±  25.2 ms    [User: 353.4 ms, System: 39.3 ms]
  Range (min … max):   389.2 ms … 501.5 ms    20 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  ~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v ran
    1.04 ± 0.07 times faster than ~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v
    1.43 ± 0.11 times faster than ~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v
```

`5.2.0` and `5.2.0+pr13086` both have similar performance and are much faster than `4.14.1`.

**Comment by gasche** [comment]
Two questions:
- can you have a look at the `perf` profile to see if you see `do_some_marking` work with the 5.2 version, and with the 4.14 version?
- can you ensure that you don't have an OCAMLRUNPARAM variable set in your environment? Coq checks for it and cancels its own gc-setting changes when it exists, and affects the 4.14 result (which are faster with the Coq-default GC settings than with the OCaml-default GC settings).

**Comment by ejgallego** [comment]
@kayceesrk , interesting, I am having trouble reproducing the results on my intel laptop too.

The way we worked was:
- we started from the numbers from the Coq bench (c.f. https://github.com/coq/coq/pull/19225)
- looking at these numbers in more detail, we suspected a regression on marshall
- we used @gasche machine (AMD?) to reproduce these perf results, with success
- @gasche and @OlivierNicole debugged marshall and created this issue
- in parallel, I tried to bench locallly on my intel machine more, as I have a version of Coq that doesn't use marshall (it preloads .vo files). My goal was to understand if marshall alone was responsible for the perf hit in Coq's bench. However, I got similar results to you.

So IMHO this issue is becoming quite tricky!

**Comment by OlivierNicole** [comment]
~~I also can’t reproduce (we worked on @gasche’s AMD machine) and get yet different results on my Intel Core i7:~~ EDIT: wrong, see below

```
$ hyperfine -L version 4,5 "coq-ocaml{version}/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml{version}/_build_ci/color/ CoLoR TestMarshal.v" 
Benchmark 1: coq-ocaml4/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml4/_build_ci/color/ CoLoR TestMarshal.v
  Time (mean ± σ):     413.8 ms ±   6.8 ms    [User: 348.7 ms, System: 61.4 ms]
  Range (min … max):   404.5 ms … 428.2 ms    10 runs
 
Benchmark 2: coq-ocaml5/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml5/_build_ci/color/ CoLoR TestMarshal.v
  Time (mean ± σ):     436.1 ms ±   4.1 ms    [User: 367.9 ms, System: 66.7 ms]
  Range (min … max):   427.9 ms … 443.2 ms    10 runs
 
Summary
  coq-ocaml4/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml4/_build_ci/color/ CoLoR TestMarshal.v ran
    1.05 ± 0.02 times faster than coq-ocaml5/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml5/_build_ci/color/ CoLoR TestMarshal.v
```

**Comment by gasche** [comment]
Okay, so my impression so far is that there is no *conclusive* evidence of a real slowdown. I suspect that the GC pacing is off when unmarshalling data, but have been unable to reproduce this robust in a synthetic setting. I propose to close and forget about this for now. We can reopen if more evidence comes up that this is an issue in practice (in particular, @ejgallego tried to remove the marshal-size accounting and rerun benchmarks on the Coq CI machines, maybe the new data will suggest an important effect in that setting).

**Comment by OlivierNicole** [comment]
Sorry, scratch what I wrote, I got owned by this:

> can you ensure that you don't have an OCAMLRUNPARAM variable set in your environment? Coq checks for it and cancels its own gc-setting changes when it exists, and affects the 4.14 result (which are faster with the Coq-default GC settings than with the OCaml-default GC settings).

Once OCAMLRUNPARAM is unset I can in fact reproduce:

```
hyperfine -L version 4,5 "coq-ocaml{version}/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml{version}/_build_ci/color/ CoLoR TestMarshal.v"
Benchmark 1: coq-ocaml4/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml4/_build_ci/color/ CoLoR TestMarshal.v
  Time (mean ± σ):     258.2 ms ±   9.7 ms    [User: 166.2 ms, System: 90.2 ms]
  Range (min … max):   251.1 ms … 284.0 ms    11 runs
 
Benchmark 2: coq-ocaml5/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml5/_build_ci/color/ CoLoR TestMarshal.v
  Time (mean ± σ):     352.2 ms ±   6.1 ms    [User: 260.9 ms, System: 88.0 ms]
  Range (min … max):   344.6 ms … 363.4 ms    10 runs
 
Summary
  coq-ocaml4/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml4/_build_ci/color/ CoLoR TestMarshal.v ran
    1.36 ± 0.06 times faster than coq-ocaml5/_build/default/topbin/coqc_bin.exe -w -all -R coq-ocaml5/_build_ci/color/ CoLoR TestMarshal.v
```

**Comment by silene** [comment]
If you want some synthetic benchmark instead of a real-world one, you could just marshal and unmarshal a long list, e.g., with the following OCaml program. (Run it first as `foo out foo.data` to produce the marshaled data.)
```ocaml
let () =
  let nb = 10_000_000 in
  if Sys.argv.(1) = "out" then
    let rec f acc n = if n = 0 then acc else f (n :: acc) (n - 1) in
    let l = f [] nb in
    let fout = open_out_bin Sys.argv.(2) in
    let () = Marshal.to_channel fout l [] in
    close_out fout
  else
    let fin = open_in_bin Sys.argv.(2) in
    let l = Marshal.from_channel fin in
    let () = assert (List.hd l = 1) in
    close_in fin
```
```console
$ hyperfine --warmup 10 -L version 4,5 "~/tmp/foo{version} in foo.data"
Benchmark 1: ~/tmp/foo4 in foo.data
  Time (mean ± σ):     183.1 ms ±   0.9 ms    [User: 55.0 ms, System: 127.9 ms]
  Range (min … max):   181.4 ms … 184.3 ms    16 runs
 
Benchmark 2: ~/tmp/foo5 in foo.data
  Time (mean ± σ):     230.4 ms ±   2.3 ms    [User: 104.1 ms, System: 126.1 ms]
  Range (min … max):   227.0 ms … 235.2 ms    12 runs
 
Summary
  '~/tmp/foo4 in foo.data' ran 1.26 ± 0.01 times faster than '~/tmp/foo5 in foo.data'
```

Note that, if we disregard the system time (which is similar for both versions), then the unmarshal from OCaml 4 is actually twice as fast as the one from OCaml 5 on this micro-benchmark. Also, `perf` does not complain about `do_some_marking` here.
```
  87,87%  foo4     foo4                  [.] intern_rec
---
  42,49%  foo5     foo5                  [.] intern_rec
  38,94%  foo5     foo5                  [.] caml_shared_try_alloc
  11,99%  foo5     foo5                  [.] intern_alloc_obj
```

**Comment by gasche** [comment]
I believe that this is a different performance difference than the one I am reporting here, precisely because there is no major-GC work involved in the slowdown. The slowdown you observe comes from the move to `caml_shared_try_alloc`, for which long lists are basically a worst-case (lots of small objects).

**Comment by kayceesrk** [comment]
I can confirm that the results on an Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz match what @gasche, @OlivierNicole, and @ejgallego find — 4.14 is ~40% faster than 5.2. It's odd that M2 has an almost similar inverse relationship in performance. 

```bash
hyperfine -r 20 --warmup 5 '~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v' '~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v' '~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v'
Benchmark 1: ~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     578.6 ms ±   8.2 ms    [User: 385.6 ms, System: 192.7 ms]
  Range (min … max):   563.3 ms … 592.9 ms    20 runs
 
Benchmark 2: ~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     797.4 ms ±  38.3 ms    [User: 602.3 ms, System: 194.8 ms]
  Range (min … max):   767.5 ms … 945.5 ms    20 runs
 
Benchmark 3: ~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     805.7 ms ±  32.0 ms    [User: 607.2 ms, System: 198.3 ms]
  Range (min … max):   782.0 ms … 930.2 ms    20 runs
 
Summary
  '~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v' ran
    1.38 ± 0.07 times faster than '~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v'
    1.39 ± 0.06 times faster than '~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v'
```

**Comment by kayceesrk** [comment]
I wonder whether this may be a case of Coq's GC settings not being the right one for OCaml 5. I already see a specialisation for best fit collector. 

https://github.com/coq/coq/blob/c8c33bb72425cd9427d99a9d120a863cee09647c/sysinit/coqinit.ml#L52-L76

If I set OCAMLRUNPARAM, then all the variants perform similarly on x86. 

```
$ export OCAMLRUNPARAM="x=1" # x is not one of the options handled by OCaml and is ignored. 
$ hyperfine -r 20 --warmup 5 '~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v' '~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v' '~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v'
Benchmark 1: ~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     850.4 ms ±  12.5 ms    [User: 700.3 ms, System: 149.9 ms]
  Range (min … max):   823.2 ms … 864.0 ms    20 runs
 
Benchmark 2: ~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     829.2 ms ±  16.7 ms    [User: 687.5 ms, System: 141.4 ms]
  Range (min … max):   795.2 ms … 853.4 ms    20 runs
 
Benchmark 3: ~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     831.2 ms ±  22.6 ms    [User: 685.5 ms, System: 145.4 ms]
  Range (min … max):   752.4 ms … 857.1 ms    20 runs
 
Summary
  '~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v' ran
    1.00 ± 0.03 times faster than '~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v'
    1.03 ± 0.03 times faster than '~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v'
```

OTOH, on macOS, OCaml 5.2 remains faster than OCaml 4.14.1.

```
% export OCAMLRUNPARAM="x=1" # x is not one of the options handled by OCaml and is ignored. 
% hyperfine -r 20 --warmup 5 '~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v' '~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v' '~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v'
Benchmark 1: ~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     578.7 ms ±  67.1 ms    [User: 515.7 ms, System: 42.6 ms]
  Range (min … max):   479.1 ms … 759.7 ms    20 runs
 
Benchmark 2: ~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     390.6 ms ±   8.5 ms    [User: 344.8 ms, System: 39.3 ms]
  Range (min … max):   379.3 ms … 410.1 ms    20 runs
 
Benchmark 3: ~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v
  Time (mean ± σ):     417.0 ms ±  37.4 ms    [User: 360.6 ms, System: 41.8 ms]
  Range (min … max):   391.1 ms … 508.5 ms    20 runs
 
Summary
  ~/.opam/5.2.0/bin/coqc -w -all Lterm_imports.v ran
    1.07 ± 0.10 times faster than ~/repos/ocaml5.2+13086/_opam/bin/coqc -w -all Lterm_imports.v
    1.48 ± 0.17 times faster than ~/.opam/4.14.1/bin/coqc -w -all Lterm_imports.v
```

**Comment by gasche** [comment]
We studied the OCAMLRUNPARAM impact rather carefully with @OlivierNicole on Monday. We believe that the slowdown is not related by OCAMLRUNPARAM, in the following sense:

- with the OCaml-default GC settings, both OCaml 4 and OCaml 5 are somewhat slow, and their performance is comparable
- with the Coq-default GC settings, OCaml 4 is noticeably faster than with the OCaml-default settings; this shows that the Coq people were right to tune the GC settings (for this specific testcase at least)
- changing the GC settings does *not* improve the performance of this program on OCaml 5 (at least for `space_overhead` between `80` and `300`), so it remains slow, even for larger values that makes OCaml 4 faster

For OCaml 4, with the OCaml-default GC settings (o=120 in particular), we do see `do_some_marking` high in the profile. This shows that the pacing issues that slow down the OCaml 5 version also show up for some space-overhead settings -- which is reasonable, given that lower space-overhead forces the GC to collect more. But when we set the space-overhead value to larger yet reasonable values, like o=200 as is done by Coq by default, then this marking work disappears on OCaml 4, but it remains on OCaml 5. (Same with o=300; we didn't try higher values as those would be rather unreasonable for other Coq workloads.)

Note: you can play with OCAMLRUNPARAM settings easily with `hyperfine`: `hyperfine -L overhead 80,120,160,200,240 -L version 4.14.1,5.2.0 "OCAMLRUNPARAM='s=256M,o={overhead}' ..."`. On my machine I see that OCaml 5 versions are slower than OCaml 4 versions.

**Comment by kayceesrk** [comment]
Ok. I agree that it is worth a look beyond attributing the behaviour to coq's default gc settings. And, OCaml 5 is much faster on the M2 MacBook with both the OCaml default and Coq default GC settings. Would be good to understand what's going on here.

**Comment by gasche** [comment]
If marking is the bottleneck, this could be a matter of memory bandwidth?

**Comment by gasche** [comment]
@ejgallego ran more Coq benchmarks, including a benchmark with my patch applied (the one-liner mentioned above which disables work-accounting for the unmarshaller). The results could suffer from measurement biases, but they seem to suggest that the patch *does* speedup Coq noticeably for a variety of packages, beyond the short testcase I proposed above.

Benchmark results before the patch (comparing 4.14 and vanilla 5.2.0): https://github.com/coq/coq/pull/19225#issuecomment-2229965568

Benchmark results after the patch (comparing 4.14 and patched 5.2.0): https://github.com/coq/coq/pull/19225#issuecomment-2232215271

Before the patch, the performance impact of moving from 4.14 to 5.2 ranges between -1.5% (a small speedup) to +22% (a large slowdown) for coq-fiat-core; the 22% slowdown is an outlier, the second worst is 12% and there is a continuous spread between -2% and 12%. Out of the 40 tested packages, 25 have a slowdown of 5% or worse.

After the patch, the performance impact of moving from 4.14 to patched-5.2 ranges between -6.5% (a noticeable speedup) and 10%, with a continuous spread. Out of the 40 tested packages, only 7 have a slowdown of 5% or worse.

As @kayceesrk suggested, it would be nice to test the impact of #13086 (a trunk-only GC pacing change/fix) on the testcase -- my intuition is that it will change nothing, because it does not change the accounting of unmarshalling, but it might improve things due to other parts having changed. Even if it changes nothing, it is probably easier to write a "nice" fix for this pacing issue on top of it -- my idea is to keep the increment of `allocated_words` (unlike in the current one-liner, which is suspicious), but remove the increment of `allocated_words_direct`, or maybe increment it only for blocks of size above Max_young_wosize (that would go directly in the major heap if they were allocated outside marshalling).

**Comment by kayceesrk** [comment]
> As @kayceesrk suggested, it would be nice to test the impact of https://github.com/ocaml/ocaml/pull/13086 (a trunk-only GC pacing change/fix) on the testcase -- my intuition is that it will change nothing, because it does not change the accounting of unmarshalling, but it might improve things due to other parts having changed.

My experiments show that nothing significant changes.

**Comment by OlivierNicole** [comment]
In discussions, @dra27 emitted the idea that the same amount of marking may be done in 4.14 and 5.2, but locality of data became worse in 5.2 due to the separation of pools by size, which could explain the longer time spent in marking.

This is under the hypothesis that our current “fix” does not remove the slowdown but only delays marking to the point that the program stops before beginning the next GC cycle. And so we have the impression that the program is faster, but it wouldn’t be if it ran longer with more GC cycles.

**Comment by NickBarnes** [comment]
In today's runtime/GC meeting, we discussed this issue. It seems to be an example of a ["pig-in-the-python"](https://www.memorymanagement.org/glossary/p.html#term-pig-in-the-python): allocation of a lot of long-lived data causing the GC to do more work than necessary. As it happens we allocate this data directly on the major heap, so the additional work is not the nursery collections and promotions usually associated with pigs in pythons, but it's still recognisably the same issue.
This raises the possibility of a general mechanism for letting the GC know that we're allocating a pig, allowing it to adjust its accounting and not over-work. Imagine defining `caml_pig_allocation_start()` and `caml_pig_allocation_stop()` (not good names), which would switch the GC's accounting (for the current domain) between two modes (in fact you'd want some sort of "pig depth"). From Caml we could have `GC.pig : ('a -> 'b) -> 'a -> 'b`.

**Comment by NickBarnes** [comment]
There are other allocation patterns which the GC might benefit from knowing about, such as ["ramps"](https://memory-pool-system.readthedocs.io/en/latest/topic/pattern.html). Maybe we ought to have a generic mechanism.

**Comment by ejgallego** [comment]
Thanks for the detailed analysis folks; indeed @gasche patch helped quite a bit in the Coq bench; I am still stuck as I cannot observe any slowdown on my i7-10875H CPU, which as of today is the only CPU I have at hand for benchmarking.

**Comment by OlivierNicole** [comment]
Did you check that OCAMLRUNPARAM is unset?

**Comment by kayceesrk** [comment]
I confirm that the "patch" that @gasche improves performance on my x86 machine. The traces produced using `olly` shows that there is improvement

Trace for 5.2:
<img width="1691" alt="image" src="https://github.com/user-attachments/assets/d09706e2-bd67-4974-9622-c900d427425c">

Trace for 5.2+patch:
<img width="1692" alt="image" src="https://github.com/user-attachments/assets/7cc8d9eb-2b6a-4fef-8338-f60c74ee3ee8">


However, it is still unclear to me whether the GC should be implicated. A useful question to ask is why 4.14 doesn't suffer from the sped up GC.

**Comment by gasche** [comment]
In case it can help, here is my subjective perception of the current state of this PR: I stopped looking at the 4.14 behavior (I understand that it suffers from the same issue, but it is less visible due to other implementation choices that we cannot hope to replicate for OCaml 5), and I am waiting for feedback from people who know more about the GC pacing, in particular @damiendoligez, but possibly also @NickBarnes (who recognised the issue in the first place, if I understand things correctly) and @stedolan.

I would like to know if we want to merge this PR or an improvement of it, or if we want something more flexible/general/tweakable from the start.

**Comment by gasche** [comment]
@NickBarnes : @damiendoligez mentioned that you had ideas for an API to declare "ramp up" phases in memory allocation, which may have involved a higher-order wrapper to delineate ramp-up phases. Would you mind sharing a sketch of your ideas here or somewhere else? I'm working on unmarshalling-specifc allocation policy settings in the context of #13320, but I'm unconvinced by the resulting API noise and I wonder setting a Caml_state field would result in nicer code.

For now my idea would be to have:

- a per-domain `policy` flag in `Caml_state`, which could be either STEADY (the default / common case) or RAMP_UP
- a higher-order wrapper `Gc.with_ramp_up : (unit -> 'a) -> deferred_work * 'a`, which temporarily switches the current domaine to the RAMP_UP policy but keeps track of the GC work that was deferred 
- a collection function `Gc.perform_deferred_work : deferred_work -> unit`, which re-incorporates the deferred work into the GC work calculation (the idea is that users are encouraged to call this function at the place where they believe the ramp-up memory is not used anymore, if they want the corresponding amount of memory to be collected)

**Comment by NickBarnes** [comment]
I quite like your `deferred_work` idea (although maybe not the name: naming is hard). Some thought will be required about the context of a `policy` flag: is it (or should it be) thread-local? fibre-local? Maybe we need to expose more of the GC policy/pacing mechanisms and data to Caml code, so that control-flow libraries can have more say in scheduling GC work?

**Comment by gasche** [comment]
When calling `Gc.with_ramp_up f`, what we ideally want is for all code in the dynamic extent of the `f ()` invocation to run under this mode. Unfortunately we don't track this relation in the runtime (GHC tries to, see [Rules for attributing costs (to cost centres)](https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#rules-for-attributing-costs)). This is not the same as fibre-local, at least it should be inherited by forked fibers.

I'm trying to find out what would be an acceptable first (mergeable) iteration for this idea, to make progress on my specific use-case of the fixing the 5.x Coq unmarshalling performance regression. In this context my impression is that making this state domain-local is a good first approximation, in line with how we track GC statistics.

**Comment by WardBrian** [comment]
I have another organic example of a place this regression showed up in my own code (https://github.com/stan-dev/stanc3/pull/1498)

Small benchmark (i.e., maximum impact due to overhead):
```sh
$ hyperfine --warmup 20 -L version 4.14,5.3 "./stanc-marshal-{version} test/integration/good/code-gen/8_schools_ncp.stan"
Benchmark 1: ./stanc-marshal-4.14 test/integration/good/code-gen/8_schools_ncp.stan
  Time (mean ± σ):      19.0 ms ±   1.0 ms    [User: 11.5 ms, System: 7.3 ms]
  Range (min … max):    16.9 ms …  23.0 ms    167 runs
 
Benchmark 2: ./stanc-marshal-5.3 test/integration/good/code-gen/8_schools_ncp.stan
  Time (mean ± σ):      27.6 ms ±   1.6 ms    [User: 19.4 ms, System: 8.1 ms]
  Range (min … max):    25.1 ms …  37.0 ms    114 runs
 
Summary
  ./stanc-marshal-4.14 test/integration/good/code-gen/8_schools_ncp.stan ran
    1.46 ± 0.12 times faster than ./stanc-marshal-5.3 test/integration/good/code-gen/8_schools_ncp.stan
```

Larger:

```sh
$ hyperfine --warmup 20 -L version 4.14,5.3 "./stanc-marshal-{version} test/integration/good/code-gen/mother.stan"
Benchmark 1: ./stanc-marshal-4.14 test/integration/good/code-gen/mother.stan
  Time (mean ± σ):      62.9 ms ±   2.0 ms    [User: 52.3 ms, System: 10.3 ms]
  Range (min … max):    57.9 ms …  66.9 ms    47 runs
 
Benchmark 2: ./stanc-marshal-5.3 test/integration/good/code-gen/mother.stan
  Time (mean ± σ):      72.3 ms ±   3.8 ms    [User: 61.2 ms, System: 10.7 ms]
  Range (min … max):    67.5 ms …  88.8 ms    43 runs
 
Summary
  ./stanc-marshal-4.14 test/integration/good/code-gen/mother.stan ran
    1.15 ± 0.07 times faster than ./stanc-marshal-5.3 test/integration/good/code-gen/mother.stan
```

**Comment by gasche** [comment]
#13861 should have fixed this issue, in so far as developers can modify the affected code by calling `Gc.ramp_up` around the costly unmarshalling call, and possibly `Gc.ramp_down` when they believe that this memory is going to be released -- or not at all, if the memory is likely to stay alive for a long time.
